File: app/src/main/java/com/akari/ppx/xp/hook/code/purity/VideoHook.java
package com.akari.ppx.xp.hook.code.purity;

import android.content.Context;

import com.akari.ppx.xp.hook.code.SuperbHook;

import java.util.ArrayList;
import java.util.concurrent.FutureTask;

import de.robv.android.xposed.XC_MethodHook;

import static com.akari.ppx.common.constant.Prefs.SAVE_VIDEO;
import static de.robv.android.xposed.XposedHelpers.callMethod;

public class VideoHook extends SuperbHook {
	@Override
	protected void onHook(ClassLoader cl) {
		hookMethod(SAVE_VIDEO, "com.sup.android.video.VideoDownloadHelper", "doDownload", Context.class, "com.sup.android.base.model.VideoModel"
				, "com.sup.android.video.VideoDownLoadConfig", "com.ss.android.socialbase.downloader.depend.IDownloadListener", boolean.class, "kotlin.jvm.functions.Function1", new XC_MethodHook() {
					@Override
					protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
						Object videoModel = param.args[1];
						FutureTask<String> task = new FutureTask<>(new WebTask(2, (String) callMethod(videoModel, "getUri"), cl));
						new Thread(task).start();
						callMethod(((ArrayList) callMethod(videoModel, "getUrlList")).get(0), "setUrl", task.get());
					}
				});
	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/purity/VideoHook.kt
@file:Suppress("unused")

package com.akari.ppx.xp.hook.purity

import android.content.Context
import android.util.Base64
import com.google.gson.JsonElement
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import com.akari.ppx.utils.*
import com.akari.ppx.xp.Init.cl
import com.akari.ppx.xp.hook.SwitchHook

class VideoHook : SwitchHook("save_video") {
    override fun onHook() {
        "com.sup.android.video.VideoDownloadHelper".findClass(cl).apply {
            replaceMethod(
                "doDownload",
                Context::class.java,
                "com.sup.android.base.model.VideoModel",
                "com.sup.android.video.VideoDownLoadConfig",
                "com.ss.android.socialbase.downloader.depend.IDownloadListener",
                Boolean::class.java,
                "kotlin.jvm.functions.Function1"
            ) { param ->
                CoroutineScope(Dispatchers.Main).launch {
                    val videoModel = param.args[1]
                    var resp: String

                    fun String.get() = String(
                        "com.bytedance.apm.net.DefaultHttpServiceImpl".findClass(cl).new()
                            .callMethod("doGet", this, null)?.callMethodAs<ByteArray>("b")!!
                    )

                    withContext(Dispatchers.IO) {
                        var uri = videoModel.callMethodAs<String>("getUri")
                        uri.check("") {
                            uri = "https://h5.pipix.com/bds/webapi/item/detail/?item_id=${
                                param.args[2].callMethodAs<Long>("getItemId")
                            }".get()
                                .fromJsonElement()["data"]["item"]["origin_video_id"].asString
                        }
                        val ts = System.currentTimeMillis()
                        resp = "https://i.snssdk.com/video/play/1/bds/$ts/${
                            "com.bytedance.common.utility.DigestUtils".findClass(cl)
                                .callStaticMethod(
                                    "md5Hex",
                                    "ts${ts}userbdsversion1video${uri}vtypemp4f425df23905d4ee38685e276072faa0c"
                                )
                        }/mp4/$uri".get()
                    }

                    fun JsonElement.getMaxQualityVideo(i: Int = 5): String =
                        runCatching {
                            this["video_$i"]["main_url"].asString
                        }.getOrElse { getMaxQualityVideo(i - 1) }

                    String(
                        Base64.decode(
                            resp.fromJsonElement()["video_info"]["data"]["video_list"].getMaxQualityVideo(),
                            0
                        )
                    ).let { url ->
                        videoModel.callMethodAs<ArrayList<*>>("getUrlList").forEach {
                            it.callMethod("setUrl", url)
                        }
                    }
                    param.invokeOriginalMethod()
                }
            }
            replaceMethod("isEnableDownloadGodVideo") { false }
        }
    }
}